File: app/src/test/java/com/amaze/filemanager/asynchronous/services/ftp/FtpServiceTest.java
package com.amaze.filemanager.asynchronous.services.ftp;

import android.content.Intent;
import android.content.SharedPreferences;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.net.Uri;
import android.net.wifi.WifiManager;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.util.Base64;

import com.amaze.filemanager.BuildConfig;
import com.amaze.filemanager.test.ShadowContentResolverWithLocalOutputStreamSupport;

import org.apache.commons.compress.utils.IOUtils;
import org.apache.commons.net.ftp.FTP;
import org.apache.commons.net.ftp.FTPClient;
import org.apache.commons.net.ftp.FTPFile;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.Robolectric;
import org.robolectric.RobolectricTestRunner;
import org.robolectric.RuntimeEnvironment;
import org.robolectric.Shadows;
import org.robolectric.annotation.Config;
import org.robolectric.shadow.api.Shadow;
import org.robolectric.shadows.ShadowConnectivityManager;
import org.robolectric.shadows.ShadowNetworkInfo;
import org.robolectric.shadows.ShadowWifiManager;
import org.robolectric.shadows.multidex.ShadowMultiDex;
import org.robolectric.util.ReflectionHelpers;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.FileOutputStream;
import java.io.FileWriter;
import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.Writer;
import java.net.InetAddress;
import java.net.InetSocketAddress;
import java.net.Socket;
import java.net.SocketException;
import java.nio.charset.Charset;
import java.security.SecureRandom;
import java.util.Locale;

import static com.amaze.filemanager.asynchronous.services.ftp.FtpService.KEY_PREFERENCE_PATH;
import static com.amaze.filemanager.asynchronous.services.ftp.FtpService.KEY_PREFERENCE_PORT;
import static org.junit.Assert.*;

@RunWith(RobolectricTestRunner.class)
@Config(constants = BuildConfig.class, shadows = {ShadowMultiDex.class, ShadowContentResolverWithLocalOutputStreamSupport.class})
public class FtpServiceTest {

    private static final int FTP_PORT = 62222;

    private FTPClient ftpClient;

    private FtpService service;

    @Before
    public void setUp() throws Exception {

        ShadowConnectivityManager cm = Shadows.shadowOf(RuntimeEnvironment.application.getSystemService(ConnectivityManager.class));
        ShadowWifiManager wifiManager = Shadows.shadowOf(RuntimeEnvironment.application.getSystemService(WifiManager.class));
        cm.setActiveNetworkInfo(ShadowNetworkInfo.newInstance(NetworkInfo.DetailedState.CONNECTED, ConnectivityManager.TYPE_WIFI, -1, true, true));
        wifiManager.setWifiEnabled(true);
        ReflectionHelpers.callInstanceMethod(wifiManager.getConnectionInfo(), "setInetAddress", ReflectionHelpers.ClassParameter.from(InetAddress.class, InetAddress.getLoopbackAddress()));

        service = Robolectric.setupService(FtpService.class);
        SharedPreferences prefs = PreferenceManager.getDefaultSharedPreferences(RuntimeEnvironment.application);
        prefs.edit()
            .putString(KEY_PREFERENCE_PATH, Environment.getExternalStorageDirectory().getAbsolutePath())
            .putInt(KEY_PREFERENCE_PORT, FTP_PORT).commit();
        service.onStartCommand(new Intent(FtpService.ACTION_START_FTPSERVER).putExtra(FtpService.TAG_STARTED_BY_TILE, false), 0, 0);

        assertTrue(FtpService.isRunning());
        waitForServer();

        //ShadowLog.stream = System.out;
    }

    @After
    public void tearDown() throws Exception {

        if (ftpClient != null && ftpClient.isConnected()) {
            ftpClient.logout();
            ftpClient.disconnect();
            ftpClient = null;
        }
        service.onDestroy();
    }

    @Test
    public void testPwdRoot() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
    }

    @Test
    public void testCdPastRoot() throws Exception {
        login();
        ftpClient.changeWorkingDirectory("/");
        assertEquals("/", ftpClient.printWorkingDirectory());
        ftpClient.changeWorkingDirectory("../");
        assertEquals("/", ftpClient.printWorkingDirectory());
        ftpClient.changeWorkingDirectory("../../../../../../../");
        assertEquals("/", ftpClient.printWorkingDirectory());
    }

    @Test
    public void testMkdir() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.makeDirectory("/Documents"));
        assertTrue(ftpClient.changeWorkingDirectory("/Documents"));
        assertEquals("/Documents", ftpClient.printWorkingDirectory());
        assertTrue(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
    }

    @Test
    public void testMkdirWithoutSlashPrefix() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.makeDirectory("Documents"));
        assertTrue(ftpClient.changeWorkingDirectory("Documents"));
        assertEquals("/Documents", ftpClient.printWorkingDirectory());
        assertTrue(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
    }

    @Test
    public void testMkdirPastRoot() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertFalse(ftpClient.makeDirectory("../../../../../../../../../../Documents"));
        assertFalse(ftpClient.changeWorkingDirectory("../../../../../../../../../../Documents"));
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertFalse(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
    }

    @Test
    public void testRepeatMkdir() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.makeDirectory("/Documents"));        assertTrue(ftpClient.changeWorkingDirectory("/Documents"));
        assertEquals("/Documents", ftpClient.printWorkingDirectory());
        assertTrue(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
        assertFalse(ftpClient.makeDirectory("/Documents"));
        assertFalse(ftpClient.makeDirectory("/Documents"));
    }

    @Test
    public void testRmdir() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.makeDirectory("/Documents"));
        assertTrue(ftpClient.changeWorkingDirectory("/Documents"));
        assertEquals("/Documents", ftpClient.printWorkingDirectory());
        assertTrue(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
        assertTrue(ftpClient.changeToParentDirectory());
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.removeDirectory("/Documents"));
        assertFalse(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
    }

    @Test
    public void testRmdirWithoutSlashPrefix() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.makeDirectory("Documents"));
        assertTrue(ftpClient.changeWorkingDirectory("Documents"));
        assertEquals("/Documents", ftpClient.printWorkingDirectory());
        assertTrue(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
        assertTrue(ftpClient.changeToParentDirectory());
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertTrue(ftpClient.removeDirectory("Documents"));
        assertFalse(new File(Environment.getExternalStorageDirectory(), "Documents").exists());
    }

    @Test
    public void testRmdirPastRoot() throws Exception {
        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        assertFalse(ftpClient.removeDirectory("../../../../../../../../../../Documents"));
        assertFalse(ftpClient.removeDirectory("../../../../../../../../../../../../bin"));
        assertTrue(ftpClient.changeWorkingDirectory("../../../../../../../../../.."));
        assertEquals("/", ftpClient.printWorkingDirectory());
    }

    @Test
    public void testUploadFile() throws IOException {
        byte[] bytes1 = new byte[32], bytes2 = new byte[32];
        SecureRandom sr = new SecureRandom();
        sr.setSeed(System.currentTimeMillis());
        sr.nextBytes(bytes1);
        sr.nextBytes(bytes2);

        String randomString = Base64.encodeToString(bytes1, Base64.DEFAULT);

        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        ftpClient.changeWorkingDirectory("/");
        ftpClient.enterLocalPassiveMode();
        ftpClient.setFileType(FTP.ASCII_FILE_TYPE);
        InputStream in = new ByteArrayInputStream(randomString.getBytes("utf-8"));
        ftpClient.storeFile("test.txt", in);
        in.close();
        in = new ByteArrayInputStream(bytes2);
        ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
        ftpClient.storeFile("test.bin", in);
        in.close();
        ftpClient.logout();
        ftpClient.disconnect();

        File verify = new File(Environment.getExternalStorageDirectory(), "test.txt");
        assertTrue(verify.exists());
        ByteArrayOutputStream verifyContent = new ByteArrayOutputStream();
        IOUtils.copy(new FileInputStream(verify), verifyContent);
        assertEquals(randomString, verifyContent.toString("utf-8"));
        verify.delete();
        verify = new File(Environment.getExternalStorageDirectory(), "test.bin");
        assertTrue(verify.exists());
        verifyContent = new ByteArrayOutputStream();
        IOUtils.copy(new FileInputStream(verify), verifyContent);
        assertArrayEquals(bytes2, verifyContent.toByteArray());
        verify.delete();
    }

    @Test
    public void testDownloadFile() throws IOException {
        File testFile1 = new File(Environment.getExternalStorageDirectory(), "test.txt");
        File testFile2 = new File(Environment.getExternalStorageDirectory(), "test.bin");

        byte[] bytes1 = new byte[32], bytes2 = new byte[32];
        SecureRandom sr = new SecureRandom();
        sr.setSeed(System.currentTimeMillis());
        sr.nextBytes(bytes1);
        sr.nextBytes(bytes2);

        String randomString = Base64.encodeToString(bytes1, Base64.DEFAULT);

        Writer writer = new FileWriter(testFile1);
        writer.write(randomString);
        writer.close();

        OutputStream out = new FileOutputStream(testFile2);
        out.write(bytes2, 0, bytes2.length);
        out.close();

        ShadowContentResolverWithLocalOutputStreamSupport contentResolver = Shadow.extract(RuntimeEnvironment.application.getContentResolver());
        contentResolver.registerInputStream(Uri.fromFile(testFile1), new ByteArrayInputStream(randomString.getBytes(Charset.forName("UTF-8"))));
        contentResolver.registerInputStream(Uri.fromFile(testFile2), new ByteArrayInputStream(bytes2));

        login();
        assertEquals("/", ftpClient.printWorkingDirectory());
        ftpClient.changeWorkingDirectory("/");

        FTPFile[] files = ftpClient.listFiles();
        assertEquals(2, files.length);
        for(FTPFile file:files){
            System.err.println(String.format(Locale.US, "%s %d", file.getName(), file.getSize()));
        }

        ftpClient.enterLocalPassiveMode();
        ftpClient.setFileType(FTP.ASCII_FILE_TYPE);
        ByteArrayOutputStream verify = new ByteArrayOutputStream();
        ftpClient.retrieveFile("test.txt", verify);
        verify.close();
        assertTrue(verify.size() > 0);
        assertEquals(randomString, verify.toString("utf-8"));

        ftpClient.setFileType(FTP.BINARY_FILE_TYPE);
        verify = new ByteArrayOutputStream();
        ftpClient.retrieveFile("test.bin", verify);
        verify.close();
        assertArrayEquals(bytes2, verify.toByteArray());

        ftpClient.logout();
        ftpClient.disconnect();

        testFile1.delete();
        testFile2.delete();
    }

    private void login() throws IOException {
        this.ftpClient = new FTPClient();
        ftpClient.connect("localhost", FTP_PORT);
        ftpClient.login("anonymous", "test@example.com");
    }

    //Required since FTPService is running on separate thread...
    private void waitForServer() throws IOException {
        boolean available = false;
        while(!available) {
            Socket socket = new Socket();
            try {
                socket.connect(new InetSocketAddress(InetAddress.getLocalHost(), FTP_PORT));
                socket.close();
                available = true;
            } catch(SocketException e) {
                available = false;
            }
        }
        return;
    }
}

File: app/src/test/java/com/amaze/filemanager/asynchronous/services/FtpServiceTest.kt
/*
 * Copyright (C) 2014-2021 Arpit Khurana <arpitkh96@gmail.com>, Vishal Nehra <vishalmeham2@gmail.com>,
 * Emmanuel Messulam<emmanuelbendavid@gmail.com>, Raymond Lai <airwave209gt at gmail.com> and Contributors.
 *
 * This file is part of Amaze File Manager.
 *
 * Amaze File Manager is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */

package com.amaze.filemanager.asynchronous.services

import android.content.Context
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.net.Uri
import android.net.wifi.WifiInfo
import android.net.wifi.WifiManager
import android.os.Build.VERSION_CODES.KITKAT
import android.os.Environment
import androidx.preference.PreferenceManager
import androidx.test.core.app.ApplicationProvider
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.amaze.filemanager.asynchronous.services.ftp.FtpService
import com.amaze.filemanager.filesystem.ftpserver.AndroidFileSystemFactory
import com.amaze.filemanager.shadows.ShadowMultiDex
import org.apache.commons.net.ftp.FTPClient
import org.apache.ftpserver.ConnectionConfigFactory
import org.apache.ftpserver.FtpServer
import org.apache.ftpserver.FtpServerFactory
import org.apache.ftpserver.listener.ListenerFactory
import org.apache.ftpserver.usermanager.impl.BaseUser
import org.junit.After
import org.junit.Assert.*
import org.junit.Before
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.Shadows.shadowOf
import org.robolectric.annotation.Config
import org.robolectric.annotation.LooperMode
import org.robolectric.shadows.ShadowNetworkInfo
import org.robolectric.util.ReflectionHelpers
import java.io.ByteArrayInputStream
import java.io.File
import java.io.FileInputStream
import java.io.FileOutputStream
import java.net.InetAddress
import kotlin.random.Random

@RunWith(AndroidJUnit4::class)
@Config(sdk = [KITKAT], shadows = [ShadowMultiDex::class])
@LooperMode(LooperMode.Mode.PAUSED)
@Suppress("StringLiteralDuplication")
class FtpServiceTest {

    private val FTP_PORT = 62222

    private var server: FtpServer? = null

    private val randomContent = Random.nextBytes(16)

    private var ftpClient: FTPClient? = null

    companion object {

        val directories = arrayOf(
            Environment.DIRECTORY_MUSIC,
            Environment.DIRECTORY_PODCASTS,
            Environment.DIRECTORY_RINGTONES,
            Environment.DIRECTORY_ALARMS,
            Environment.DIRECTORY_NOTIFICATIONS,
            Environment.DIRECTORY_PICTURES,
            Environment.DIRECTORY_MOVIES,
            Environment.DIRECTORY_DOWNLOADS,
            Environment.DIRECTORY_DCIM,
            Environment.DIRECTORY_DOCUMENTS,
            "1/2/3/4/5/6/7"
        )
    }

    /**
     * Test setup
     */
    @Before
    fun setUp() {
        Environment.getExternalStorageDirectory().run {
            directories.forEach { dir ->
                File(this, dir).mkdirs()
            }
        }

        setupNetwork()
        PreferenceManager.getDefaultSharedPreferences(ApplicationProvider.getApplicationContext())
            .run {
                edit().putString(
                    FtpService.KEY_PREFERENCE_PATH,
                    Environment.getExternalStorageDirectory().absolutePath
                )
                    .apply()
            }

        File(Environment.getExternalStorageDirectory(), "test.bin").let { file ->
            file.writeBytes(randomContent)
            shadowOf(ApplicationProvider.getApplicationContext<Context>().contentResolver)
                .registerInputStream(Uri.fromFile(file), FileInputStream(file))
        }

        FtpServerFactory().run {
            val connectionConfigFactory = ConnectionConfigFactory()
            val user = BaseUser()
            user.name = "anonymous"
            user.homeDirectory = Environment.getExternalStorageDirectory().absolutePath
            connectionConfigFactory.isAnonymousLoginEnabled = true
            connectionConfig = connectionConfigFactory.createConnectionConfig()
            userManager.save(user)

            fileSystem = AndroidFileSystemFactory(
                ApplicationProvider.getApplicationContext()
            )
            addListener(
                "default",
                ListenerFactory().also {
                    it.port = FTP_PORT
                }.createListener()
            )

            server = createServer().apply {
                start()
            }
        }

        ftpClient = FTPClient().also {
            it.connect("127.0.0.1", FTP_PORT)
            it.login("anonymous", "no@e.mail")
            it.enterLocalPassiveMode()
        }
    }

    private fun setupNetwork() {
        val cm = shadowOf(
            ApplicationProvider.getApplicationContext<Context>()
                .getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        )
        val wifiManager = shadowOf(
            ApplicationProvider.getApplicationContext<Context>()
                .getSystemService(Context.WIFI_SERVICE) as WifiManager
        )
        cm.setActiveNetworkInfo(
            ShadowNetworkInfo.newInstance(
                NetworkInfo.DetailedState.CONNECTED,
                ConnectivityManager.TYPE_WIFI,
                -1,
                true,
                NetworkInfo.State.CONNECTED
            )
        )
        ReflectionHelpers.callInstanceMethod<Any>(
            wifiManager,
            "setWifiEnabled",
            ReflectionHelpers.ClassParameter.from(Boolean::class.java, true)
        )
        ReflectionHelpers.callInstanceMethod<WifiInfo>(
            wifiManager,
            "getConnectionInfo"
        ).run {
            ReflectionHelpers.callInstanceMethod<Any>(
                this,
                "setInetAddress",
                ReflectionHelpers.ClassParameter.from(
                    InetAddress::class.java,
                    InetAddress.getLoopbackAddress()
                )
            )
        }
    }

    /**
     * Kill FTP server if there is one running
     */
    @After
    fun tearDown() {
        ftpClient?.logout()
        server?.stop()
    }

    /**
     * Test on change directory functions
     */
    @Test
    fun testChdir() {
        ftpClient!!.run {
            assertEquals(directories.size + 1, listFiles().size)
            assertTrue(changeWorkingDirectory("Download"))
            assertEquals(0, listFiles().size)
            assertTrue(changeWorkingDirectory(".."))
            assertEquals(directories.size + 1, listFiles().size)
            assertTrue(changeWorkingDirectory("/1/2/3/4/5/6/7"))
            assertEquals(0, listFiles().size)
            assertTrue(changeWorkingDirectory("../"))
            assertTrue(printWorkingDirectory().startsWith("/1/2/3/4/5/6"))
            assertTrue(changeToParentDirectory())
            assertTrue(printWorkingDirectory().startsWith("/1/2/3/4/5"))
            assertTrue(changeWorkingDirectory("../../.."))
            assertTrue(printWorkingDirectory().startsWith("/1/2"))
        }
    }

    /**
     * Test remove directory function
     */
    @Test
    fun testRmDir() {
        ftpClient!!.run {
            assertTrue(changeWorkingDirectory("/"))
            assertTrue(makeDirectory("foobar"))
            assertEquals(directories.size + 2, listFiles().size)
            assertTrue(removeDirectory("foobar"))
            assertEquals(directories.size + 1, listFiles().size)
            assertFalse(changeWorkingDirectory("foobar"))
            assertTrue(listFiles("/foobar").isNullOrEmpty())
        }
    }

    /**
     * Test download file
     */
    @Test
    fun testDownloadFile() {
        ftpClient!!.run {
            assertFalse(deleteFile("/nonexist.file.txt"))
            assertNull(retrieveFileStream("/not/existing/file"))
            assertFalse(
                retrieveFile(
                    "/barrier/nonexist.file.jpg",
                    FileOutputStream(File.createTempFile("notused", "output"))
                )
            )
            assertTrue(changeWorkingDirectory("/"))
            retrieveFileStream("test.bin").let {
                assertNotNull(it)
                it.close()
            }
            completePendingCommand()
            assertTrue(printWorkingDirectory() == "/")
        }
    }

    /**
     * Test upload file
     */
    @Test
    fun testUploadFile() {
        ftpClient!!.run {
            storeFileStream("/test2.bin").let {
                ByteArrayInputStream(Random.nextBytes(24)).copyTo(it)
                it.flush()
                it.close()
            }
            completePendingCommand()
            assertTrue(rename("/test2.bin", "/test2.arc.bin"))
        }
    }

    /**
     * Test working with files and folders at subfolders
     */
    @Test
    fun testUploadFileToSubDir() {
        ftpClient!!.run {
            assertTrue(makeDirectory("/CROSS OVER"))
            storeFileStream("/CROSS OVER/test3.bin").let {
                ByteArrayInputStream(Random.nextBytes(24)).copyTo(it)
                it.flush()
                it.close()
            }
            completePendingCommand()
            assertTrue(changeWorkingDirectory("/CROSS OVER"))
            listFiles().let {
                assertEquals(1, it.size)
                assertEquals("test3.bin", it[0].name)
            }
            assertTrue(makeDirectory("/CROSS OVER/multiple"))
            assertTrue(makeDirectory("/CROSS OVER/multiple/levels down"))
            assertTrue(changeWorkingDirectory("/CROSS OVER/multiple"))
            assertTrue(changeWorkingDirectory("levels down"))
            assertEquals("/CROSS OVER/multiple/levels down", printWorkingDirectory())
            assertTrue(deleteFile("/CROSS OVER/test3.bin"))
            assertTrue(changeWorkingDirectory("/CROSS OVER"))
            listFiles().let {
                assertEquals(1, it.size)
                assertEquals("multiple", it[0].name)
            }
        }
    }
}
